File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.py
class Solution:
    def evaluate(self, s: str, k: list[list[str]]) -> str:
        n=len(s)

        mp={}
        for x in k:
            mp[x[0]]=x[1]

        sb=[]
        i=0
        while i<n:
            if s[i] == '(':
                j=i+1
                while s[j] != ')':
                    j+=1

                key=s[i+1: j]
                sb.append(mp.get(key, "?"))
                i=j+1
            else:
                sb.append(s[i])
                i+=1

        return ''.join(sb)

File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.java
class Solution{
    public String evaluate(String s, List<List<String>> k){
        int n=s.length();

        HashMap<String, String> mp=new HashMap<>();
        for(List<String> x: k){
            mp.put(x.get(0), x.get(1));
        }

        StringBuilder sb=new StringBuilder();
        for(int i=0; i<n; i++){
            char ch=s.charAt(i);
            if(ch == '('){
                int j=i+1;
                while(s.charAt(j) != ')'){
                    j++;
                }

                String key = s.substring(i+1, j);
                sb.append(mp.getOrDefault(key, "?"));
                i=j;
            }
            else{
                sb.append(ch);
            }
        }

        return sb.toString();
    }
}